### PR TITLE
Memory numa stats

### PR DESCRIPTION
**Pull request by katarzyna-z**
Signed-off-by: Katarzyna Kujawa <katarzyna.kujawa@intel.com>

This pull request introduces information from memory.numa_stat as Prometheus metrics.

### CODE REVIEW

File: metrics/prometheus.go [hunk 1]
```diff
@@ -454,6 +455,38 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
...
+						[]string{"total", "container"}, s.Timestamp)...)
+					values = append(values, getNumaStatsPerNode(s.Memory.ContainerData.NumaStats.File,
+						[]string{"file", "container"}, s.Timestamp)...)
+					values = append(values, getNumaStatsPerNode(s.Memory.ContainerData.NumaStats.Anon,
+						[]string{"anon", "container"}, s.Timestamp)...)
+					values = append(values, getNumaStatsPerNode(s.Memory.ContainerData.NumaStats.Unevictable,
+						[]string{"unevictable", "container"}, s.Timestamp)...)
+
+					values = append(values, getNumaStatsPerNode(s.Memory.HierarchicalData.NumaStats.Total,
+						[]string{"total", "hierarchy"}, s.Timestamp)...)
```
Comment (dashpole) [inline metrics/prometheus.go:478]: I suspect hierarchy vs container may need to be separate metrics by the same logic above.

File: metrics/prometheus.go [hunk 2]
```diff
@@ -454,6 +455,38 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 			},
 		}...)
 	}
+	if includedMetrics.Has(container.MemoryNumaMetrics) {
+		c.containerMetrics = append(c.containerMetrics, []containerMetric{
+			{
+				name:        "container_memory_numa_pages",
+				help:        "Memory usage per numa node",
+				valueType:   prometheus.GaugeValue,
+				extraLabels: []string{"type", "scope", "node"},
```
Comment (dashpole) [inline metrics/prometheus.go:464]: I think we only add two labels to metrics below?

File: metrics/prometheus.go [hunk 3]
```diff
@@ -454,6 +455,38 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 			},
 		}...)
 	}
+	if includedMetrics.Has(container.MemoryNumaMetrics) {
+		c.containerMetrics = append(c.containerMetrics, []containerMetric{
+			{
+				name:        "container_memory_numa_pages",
```
Comment (dashpole) [inline metrics/prometheus.go:461]: is this measured in #pages, or bytes?  The help text should specify the units, and the suffix of the metric should be _<units>

File: metrics/prometheus.go [hunk 4]
```diff
@@ -454,6 +455,38 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
...
+				help:        "Memory usage per numa node",
+				valueType:   prometheus.GaugeValue,
+				extraLabels: []string{"type", "scope", "node"},
+				getValues: func(s *info.ContainerStats) metricValues {
+					values := make(metricValues, 0)
+
+					values = append(values, getNumaStatsPerNode(s.Memory.ContainerData.NumaStats.Total,
+						[]string{"total", "container"}, s.Timestamp)...)
```
[the remark this answers is not in the document]
Comment (katarzyna-z) [reply metrics/prometheus.go:469]: In [documentation](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt) it is written that "total" count is sum of file + anon + unevictable so I'll remove metrics with "total" label.

File: metrics/prometheus.go (hunk 1 above)
Comment (katarzyna-z) [reply metrics/prometheus.go:478]: I followed the pattern which was used for "container_memory_failures_total metric", see [this](https://github.com/google/cadvisor/blob/master/metrics/prometheus.go#L426). There is "scope" label with values "container" or "hierarchy".

File: metrics/prometheus.go (hunk 3 above)
Comment (katarzyna-z) [reply metrics/prometheus.go:461]: [Here](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt) it is written that memory.numa_stat contains pages and I see that in [runc](https://github.com/opencontainers/runc/blob/d65df61dc5f87e5be4459782ba7a72619f145dca/libcontainer/cgroups/fs/memory.go#L286) values are only read from file. I'll improve help text.

**Comment by katarzyna-z** [comment]
@dashpole Could you take a look if new names of metrics are better?

File: metrics/prometheus.go (hunk 2 above)
Comment (dashpole) [reply metrics/prometheus.go:464]: I think this comment is still relevant

Comment (dashpole) [reply metrics/prometheus.go:464]: Ah, I see now.

**Pull request merged**